nexus/temporal-payloads.md

```diff
@@ -0,0 +1,74 @@
+# Temporal Payloads over Nexus+HTTP
+
+## Motivation
+
+For ideal integration of Temporal applications with Nexus, Temporal users should be able to leverage Temporal Payloads
+and not have to concern themselves with Nexus implementation details.
+
+When a workflow completes, we'd like Temporal server to be able to deliver the completion result to any registered Nexus
+callbacks and the callback receiver (a Temporal server) to translate the completion to a Temporal Payload.
+
+Inherently, we need a fixed set of translation rules from Payloads to Nexus+HTTP and back. These rules should be
+executed server-side.
+
+## Background
+
+Nexus reserves the entire HTTP request and response bodies (for input and successful results) for data and uses HTTP
+headers (a mapping of string to an array of strings) for transmitting metadata.
+
+> The main motivation for this design was to support streaming and avoid requiring HTTP proxies to inspect request
+> bodies.
+
+A Temporal Payload is defined as:
+
+```proto3
+message Payload {
+    map<string,bytes> metadata = 1;
+    bytes data = 2;
+}
+```
+
+## Translation rules
+
+To unmarshal a Temporal Payload from a Nexus request or response:
+
+1. Extract all `Content-Temporal-` prefixed headers into the payload metadata using the following rules:
+   1. Ignore multiple values (not supported)
+   2. Strip the key prefix and transform header case to camel case. E.g: `Content-Temporal-Foo-Bar` becomes `fooBar`
+   3. Read the values as [data URLs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URLs)
+2. Set the `encoding` Payload metadata field as follows:
+   1. If `Content-Length` is 0 or body is empty → `binary/null`
```

Comment:
I don't think null is the same as absent. We support default/zero values for parameters that are not specified, but null will fail to deserialize to some types. I think if there is no body, it's no payload. But if there is a body with just the text `null` that's `application/json`, it can be converted to `binary/null`.

Reply:
What's the difference between no payload and null?
There's not really a concept of 0 argument operations.

Reply:
So how does one not send a payload at all?

Reply:
Nexus to Temporal goes through the user's operation handler.
The handler can specify that it expects nil input and invoke the workflow however it wants.

Reply:
Hrmm, I forget if we decided on a form that will auto-map the workflow without requiring a client call. If that exists, are we telling users that they cannot use it for parameterless workflows?

Reply:
Confirmed off-PR that the only workflow shortcuts where a user doesn't invoke the workflow themselves _requires_ a single parameter on the workflow for generic reasons. So I'm ok with the idea that every Nexus has some kind of input.

Can you show an example of a high-level Nexus operation handler for an operation with no user input? Must users type `*struct{}`?